crates/red_knot/src/semantic/types/infer.rs

```diff
@@ -49,17 +50,37 @@ pub fn infer_type_from_definitions<T>(
 ) -> QueryResult<Type>
 where
     T: Debug + Iterator<Item = Definition>,
+{
+    infer_type_from_constrained_definitions(
+        db,
+        symbol,
+        definitions.map(|definition| ConstrainedDefinition {
+            definition,
+            constraints: vec![],
+        }),
+    )
+}
+
+/// Infer type of a symbol as union of the given ConstrainedDefinitions.
+#[tracing::instrument(level = "trace", skip(db))]
+pub fn infer_type_from_constrained_definitions<T>(
+    db: &dyn SemanticDb,
+    symbol: GlobalSymbolId,
+    constrained_definitions: T,
+) -> QueryResult<Type>
+where
+    T: Debug + Iterator<Item = ConstrainedDefinition>,
```

Comment:
I think we can remove the `Debug` trait from here, even if it is useful for debugging. 

I also prefer taking `IntoIterator` for methods that are generic over an iterable because it's nicer to call


```suggestion
where
    T: IntoIterator<Item = ConstrainedDefinition>,
```

Should this method be public or can we constrain it to the current module? Same for `infer_type_from_definitions`.

Reply:
The `Debug` trait constraint was here because it's required for tracing. And switching to `IntoIterator` was also giving me trouble with tracing. I just removed the tracing; not sure we really need it on these internal functions.

And I removed the `pub` as well.